### PR TITLE
converted rocketchat-integrations coffee to js

### PR DESCRIPTION
**Pull request by ggazzo**
@RocketChat/core

### CODE REVIEW

File: packages/rocketchat-integrations/server/api/api.js [hunk 1]
```diff
@@ -0,0 +1,366 @@
+/* globals Api Meteor Restivus logger processWebhookMessage*/
+// TODO: remove globals
+const vm = Npm.require('vm');
```
Comment (rodrigok) [inline packages/rocketchat-integrations/server/api/api.js]: Use import instead of require

File: packages/rocketchat-integrations/server/api/api.js [hunk 2]
```diff
@@ -0,0 +1,366 @@
...
+
+import moment from 'moment';
+
+const compiledScripts = {};
+
+const buildSandbox = function(store) {
+	if (store == null) {
+		store = {};
+	}
```
Comment (rodrigok) [inline packages/rocketchat-integrations/server/api/api.js]: Use default values instead of this condition

File: packages/rocketchat-integrations/server/api/api.js [hunk 3]
```diff
@@ -0,0 +1,366 @@
...
+	return { store, sandbox	};
+};
+
+const getIntegrationScript = function(integration) {
+	const compiledScript = compiledScripts[integration._id];
+	if ((compiledScript != null) && +compiledScript._updatedAt === +integration._updatedAt) {
+		return compiledScript.script;
+	}
+	const script = integration.scriptCompiled;
+	let vmScript = null;
```
Comment (rodrigok) [inline packages/rocketchat-integrations/server/api/api.js]: Why did you change from `undefined` to `null`?

File: packages/rocketchat-integrations/server/api/api.js [hunk 4]
```diff
@@ -0,0 +1,366 @@
...
+		alias: this.integration.alias,
+		avatar: this.integration.avatar,
+		emoji: this.integration.emoji
+	};
+	if (this.integration.scriptEnabled === true && (this.integration.scriptCompiled != null) && this.integration.scriptCompiled.trim() !== '') {
+		let script;
+		try {
+			script = getIntegrationScript(this.integration);
+		} catch (error) {
+			const e = error;
```
Comment (rodrigok) [inline packages/rocketchat-integrations/server/api/api.js]: Why declare a new var?

File: packages/rocketchat-integrations/server/api/api.js [hunk 5]
```diff
@@ -0,0 +1,366 @@
...
+		};
+		try {
+			const sandboxItems = buildSandbox(compiledScripts[this.integration._id].store);
+			const sandbox = sandboxItems.sandbox;
+			sandbox.script = script;
+			sandbox.request = request;
+			const result = vm.runInNewContext('script.process_incoming_request({ request: request })', sandbox, {
+				timeout: 3000
+			});
+			if (result && result.console.error) {
```
Comment (rodrigok) [inline packages/rocketchat-integrations/server/api/api.js]: `result.console`?

File: packages/rocketchat-integrations/server/api/api.js [hunk 6]
```diff
@@ -0,0 +1,366 @@
...
+				return RocketChat.API.v1.failure(result.error);
+			}
+			this.bodyParams = result != null ? result.content : null;
```
Comment (rodrigok) [inline packages/rocketchat-integrations/server/api/api.js]: Simplify with `this.bodyParams = result && result.content;`

File: packages/rocketchat-integrations/server/api/api.js [hunk 7]
```diff
@@ -0,0 +1,366 @@
...
+			logger.incoming.debug('[Process Incoming Request result of Trigger', this.integration.name, ':]');
+			logger.incoming.debug('result', this.bodyParams);
+		} catch (error) {
+			const e = error;
```
Comment (rodrigok) [inline packages/rocketchat-integrations/server/api/api.js]: Why this declaration?

File: packages/rocketchat-integrations/server/api/api.js [hunk 8]
```diff
@@ -0,0 +1,366 @@
...
+		i: this.integration._id
+	};
+	try {
+		const message = processWebhookMessage(this.bodyParams, this.user, defaultValues);
+		if (_.isEmpty(message)) {
+			return RocketChat.API.v1.failure('unknown-error');
+		}
+		return RocketChat.API.v1.success();
+	} catch (error) {
+		const e = error;
```
Comment (rodrigok) [inline packages/rocketchat-integrations/server/api/api.js]: Why this declaration?

File: packages/rocketchat-integrations/server/api/api.js [hunk 9]
```diff
@@ -0,0 +1,358 @@
...
+				};
+			} catch (error) {
+				return {
+					error
+				};
+			}
+		}
+	};
+
+	Object.keys(RocketChat.models).filter(function(k) {
```
Comment (graywolf336) [inline packages/rocketchat-integrations/server/api/api.js]: Can we turn these into single line arrow functions? Would clean this up quite a bit.

File: packages/rocketchat-integrations/server/api/api.js [hunk 10]
```diff
@@ -0,0 +1,358 @@
...
+	return { store, sandbox	};
+};
+
+const getIntegrationScript = function(integration) {
+	const compiledScript = compiledScripts[integration._id];
+	if ((compiledScript != null) && +compiledScript._updatedAt === +integration._updatedAt) {
+		return compiledScript.script;
+	}
+	const script = integration.scriptCompiled;
+	const sandboxItems = buildSandbox();
```
Comment (graywolf336) [inline packages/rocketchat-integrations/server/api/api.js]: Break this out using the spread operator? `const { store, sandbox } = buildSandbox();` And then use it below, as it'd clean this up and make it very similar to how the outgoing webhook script looks as well.

File: packages/rocketchat-integrations/server/api/api.js [hunk 11]
```diff
@@ -0,0 +1,358 @@
+/* globals Api Meteor Restivus logger processWebhookMessage*/
+// TODO: remove globals
+import vm from 'vm';
+import moment from 'moment';
+
+const compiledScripts = {};
+
+const buildSandbox = function(store = {}) {
```
Comment (graywolf336) [inline packages/rocketchat-integrations/server/api/api.js]: Any reason for variables instead of named functions in this entire file?

File: packages/rocketchat-integrations/server/api/api.js [hunk 12]
```diff
@@ -0,0 +1,358 @@
...
+			return {user};
+		}
+	}
+});
+
+const createIntegration = function(options, user) {
+	logger.incoming.info('Add integration', options.name);
+	logger.incoming.debug(options);
+	Meteor.runAsUser(user._id, function() {
+		switch (options['event']) {
```
Comment (graywolf336) [inline packages/rocketchat-integrations/server/api/api.js]: Convert this to `options.event` so the assessors are like the other ones for options.

File: packages/rocketchat-integrations/server/api/api.js [hunk 13]
```diff
@@ -0,0 +1,358 @@
...
+			body: 'Service Unavailable'
+		};
+	}
+	const defaultValues = {
+		channel: this.integration.channel,
+		alias: this.integration.alias,
+		avatar: this.integration.avatar,
+		emoji: this.integration.emoji
+	};
+	if (this.integration.scriptEnabled === true && (this.integration.scriptCompiled != null) && this.integration.scriptCompiled.trim() !== '') {
```
Comment (graywolf336) [inline packages/rocketchat-integrations/server/api/api.js]: Why the extra `()`'s around the middle condition?

File: packages/rocketchat-integrations/server/api/api.js [hunk 14]
```diff
@@ -0,0 +1,358 @@
...
+			content_raw: this.request._readableState && this.request._readableState.buffer && this.request._readableState.buffer.toString(),
+			headers: this.request.headers,
+			user: {
+				_id: this.user._id,
+				name: this.user.name,
+				username: this.user.username
+			}
+		};
+		try {
+			const sandboxItems = buildSandbox(compiledScripts[this.integration._id].store);
```
Comment (graywolf336) [inline packages/rocketchat-integrations/server/api/api.js]: Same as above, use the object spread. :+1:

File: packages/rocketchat-integrations/server/api/api.js [hunk 15]
```diff
@@ -0,0 +1,358 @@
...
+	logger.incoming.info('Info integration');
+	return {
+		statusCode: 200,
+		body: {
+			success: true
+		}
+	};
+};
+
+Api.addRoute(':integrationId/:userId/:token', {
```
Comment (graywolf336) [inline packages/rocketchat-integrations/server/api/api.js]: Can the `{ authRequired: true }` be moved into the same line and not several lines, this way it looks like the api definitions for the api v1.

File: packages/rocketchat-integrations/server/api/api.js (hunk 11 above)
Comment (ggazzo) [inline packages/rocketchat-integrations/server/api/api.js]: It is the "way" that I chose ~wrong~. I'll fix it. There are one difference if I tried to overwrite the function.

```javascript
const buildSandbox = function(store = {}) {}
.
.
.
buildSandbox = anotherFunction(){}<- exception

```
